config/src/main/java/com/typesafe/config/impl/ConfigImpl.java

```diff
@@ -187,42 +188,58 @@ private static ConfigOrigin valueOrigin(String originDescription) {
         else
             return SimpleConfigOrigin.newSimple(originDescription);
     }
+    
+    private static ConfigOrigin valueOrigin(String originDescription, String[] comments) {
+    	if (originDescription == null && comments == null) 
+    		return defaultValueOrigin;
+    	if (comments == null)
+            return SimpleConfigOrigin.newSimple(originDescription);
+    	if (originDescription == null)
+    		return ((SimpleConfigOrigin) defaultValueOrigin).setComments(Arrays.asList(comments));
+    	return SimpleConfigOrigin.newSimple(originDescription).setComments(Arrays.asList(comments));
+    }
 
     /** For use ONLY by library internals, DO NOT TOUCH not guaranteed ABI */
-    public static ConfigValue fromAnyRef(Object object, String originDescription) {
+    public static ConfigValue fromAnyRef(Object object, String originDescription, String[] comments) {
+        ConfigOrigin valueOrigin = valueOrigin(originDescription, comments);
         ConfigOrigin origin = valueOrigin(originDescription);
-        return fromAnyRef(object, origin, FromMapMode.KEYS_ARE_KEYS);
+        if (object != null && object instanceof ConfigValue)
+            object = ((ConfigValue) object).unwrapped();
+        return fromAnyRef(object, valueOrigin, origin, FromMapMode.KEYS_ARE_KEYS);
     }
 
     /** For use ONLY by library internals, DO NOT TOUCH not guaranteed ABI */
     public static ConfigObject fromPathMap(
-            Map<String, ? extends Object> pathMap, String originDescription) {
+            Map<String, ? extends Object> pathMap, String originDescription, String[] comments) {
+        ConfigOrigin valueOrigin = valueOrigin(originDescription, comments);
         ConfigOrigin origin = valueOrigin(originDescription);
-        return (ConfigObject) fromAnyRef(pathMap, origin,
+        if (pathMap != null && pathMap instanceof ConfigValue)
+            pathMap = (Map<String, ? extends Object>) ((ConfigValue) pathMap).unwrapped();
+        return (ConfigObject) fromAnyRef(pathMap, valueOrigin, origin,
                 FromMapMode.KEYS_ARE_PATHS);
     }
 
-    static AbstractConfigValue fromAnyRef(Object object, ConfigOrigin origin,
+    static AbstractConfigValue fromAnyRef(Object object, ConfigOrigin valueOrigin, ConfigOrigin origin,
```

Comment:
can you explain the difference between `valueOrigin` and `origin` here? it may be simple but I am not sure about it.